File: otel-extension/src/main/java/io/scalac/mesmer/zio/ZIOExecutorAdvice.java
package io.scalac.mesmer.zio;

import net.bytebuddy.asm.Advice;
import zio.Executor;

public class ZIOExecutorAdvice {

  @Advice.OnMethodExit
  static void construction(@Advice.This Executor executor) {

    System.out.println("Imma firin' an executor: " + executor.toString());

    ZIOMetricsInstrumenter.registerExecutionMetrics(executor);
  }
}

File: otel-extension/src/main/scala/io/scalac/mesmer/otelextension/instrumentations/zio/advice/ZIOExecutorAdvice.scala
package io.scalac.mesmer.otelextension.instrumentations.zio.advice

import net.bytebuddy.asm.Advice
import zio.Executor

import io.scalac.mesmer.otelextension.instrumentations.zio.ZIOExecutorMetrics

object ZIOExecutorAdvice {
  @Advice.OnMethodExit
  def constructExecutor(@Advice.This executor: Executor): Unit =
    ZIOExecutorMetrics.registerExecutorMetrics(executor)

}
